components/proxy/src/main/java/com/hotels/styx/spi/config/SpiExtension.kt

```diff
@@ -0,0 +1,86 @@
+/*
+  Copyright (C) 2013-2022 Expedia Inc.
+
+  Licensed under the Apache License, Version 2.0 (the "License");
+  you may not use this file except in compliance with the License.
+  You may obtain a copy of the License at
+
+  http://www.apache.org/licenses/LICENSE-2.0
+
+  Unless required by applicable law or agreed to in writing, software
+  distributed under the License is distributed on an "AS IS" BASIS,
+  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+  See the License for the specific language governing permissions and
+  limitations under the License.
+ */
+package com.hotels.styx.spi.config
+
+import com.fasterxml.jackson.annotation.JsonCreator
+import com.fasterxml.jackson.annotation.JsonProperty
+import com.fasterxml.jackson.databind.DeserializationFeature
+import com.fasterxml.jackson.databind.JsonNode
+import com.fasterxml.jackson.databind.ObjectMapper
+import com.fasterxml.jackson.databind.node.NullNode
+import com.fasterxml.jackson.dataformat.yaml.YAMLFactory
+import java.io.IOException
+import java.util.Objects
+
+/**
+ * Factory/configuration block.
+ */
+class SpiExtension @JsonCreator constructor(
+    @JsonProperty("factory") factory: SpiExtensionFactory,
+    @JsonProperty("config") config: JsonNode?,
+    @JsonProperty("enabled") enabled: Boolean?
+) {
+    private val factory: SpiExtensionFactory
+    private val config: JsonNode
+    private val enabled: Boolean
+
+    init {
+        this.factory = Objects.requireNonNull(factory, "Factory attribute missing")
+        this.config = config ?: NullNode.getInstance()
+        this.enabled = enabled == null
+    }
+
+    fun factory(): SpiExtensionFactory = factory
+
+    fun config(): JsonNode = config
+
+    fun enabled(): Boolean = enabled
+
+    fun <T> config(configClass: Class<T>?): T {
+        val parser = config.traverse()
+        return try {
+            MAPPER.readValue(parser, configClass)
+        } catch (e: IOException) {
+            throw RuntimeException(e)
+        }
+    }
+
+    override fun hashCode(): Int = Objects.hashCode(factory)
```

Comment:
Using a Kotlin data class would eliminate the need to manually add `hashCode`, `equals` and `toString` functions.

Reply:
If I do change this to a data class, it complains:

"Data class primary constructor must have only property (val / var) parameters"

But, I believe I can't just make these vals in the constructor because the init block needs to set factory/config/enabled to modified values from what is passed into the constructor. Current init with instance vars defined outside of constructor:
```
private val factory: SpiExtensionFactory
private val config: JsonNode
private val enabled: Boolean

init {
        this.factory = requireNonNull(factory, "Factory attribute missing")
        this.config = config ?: getInstance()
        this.enabled = enabled == null
    }
```
Let me know if there's a better way to do this. For now, I'll keep hashCode, equals, and toString but check for potential improvements within those.

Reply:
On further examination, it looks like a `data class` doesn't fit well after all, as only one of the properties is used to make equals/hashcode. Let's leave it as a normal class then.